File: apps/mobile/modules/pose-detector/android/src/main/java/expo/modules/posedetector/PoseDetectorModule.kt
package expo.modules.posedetector

import com.mrousavy.camera.frameprocessors.FrameProcessorPluginRegistry
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class PoseDetectorModule : Module() {
  companion object {
  @Volatile
    private var isPluginRegistered = false
  }

  override fun definition() = ModuleDefinition {
    Name("PoseDetector")

    OnCreate {
      if (!isPluginRegistered) {
        FrameProcessorPluginRegistry.addFrameProcessorPlugin("detectPose") { _, _ ->
          PoseFrameProcessorPlugin(appContext)
        }
        isPluginRegistered = true
      }
    }

    AsyncFunction("detect") { imageBytes: ByteArray ->
      val context = appContext.reactContext ?: return@AsyncFunction null
      PoseLandmarkerHelper.detect(context, imageBytes)
    }
  }
}

File: apps/mobile/modules/pose-detector/ios/PoseDetectorModule.swift
import ExpoModulesCore
import UIKit
import VisionCamera

public class PoseDetectorModule: Module {
  private static var isPluginRegistered = false

  public func definition() -> ModuleDefinition {
    Name("PoseDetector")

    OnCreate {
      if !PoseDetectorModule.isPluginRegistered {
        FrameProcessorPluginRegistry.addFrameProcessorPlugin("detectPose") { proxy, options in
          PoseFrameProcessorPlugin(proxy: proxy, options: options)
        }
        PoseDetectorModule.isPluginRegistered = true
      }
    }

    Function("isReady") {
      return PoseLandmarkerHelper.shared.isReady()
    }

    AsyncFunction("detect") { (imageData: Data) -> [NSNumber]? in
      guard let image = UIImage(data: imageData) else { return nil }
      return PoseLandmarkerHelper.shared.detect(image: image)
    }
  }
}
